modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Payload.scala

```diff
@@ -14,65 +14,76 @@ package com.snowplowanalytics.snowplow.scalatracker
 
 import java.util.Base64
 
-import io.circe.Json
-
-import Emitter.EmitterPayload
+import io.circe._
+import io.circe.syntax._
+import com.snowplowanalytics.iglu.core.SelfDescribingData
+import com.snowplowanalytics.iglu.core.circe.implicits._
 
 /**
  * Contains the map of key-value pairs making up an event
- * Must be used within single function as **not thread-safe**
  */
-private[scalatracker] final case class Payload(private val nvPairs: Map[String, String] = Map.empty) extends AnyVal {
+case class Payload(toMap: Map[String, String] = Map.empty) extends AnyVal
 
-  /**
-   * Add a key-value pair
-   *
-   * @param name parameter name
-   * @param value parameter value
-   */
-  def add(name: String, value: String): Payload =
-    addDict(Map(name -> value))
+object Payload {
 
-  /**
-   * Overloaded add function for Option. Don't modify payload for None
-   *
-   * @param name parameter name
-   * @param value optional parameter value
-   */
-  def add(name: String, value: Option[String]): Payload =
-    value match {
-      case Some(v) => add(name, v)
-      case None    => this
-    }
+  implicit val payloadEncoder: Encoder[Payload] =
+    Encoder.encodeMap[String, String].contramap[Payload](_.toMap)
 
-  /** Add a map of key-value pairs one by one */
-  def addDict(dict: Map[String, String]): Payload = {
-    val filtered = dict.filter { case (key, value) => key != null && value != null && !key.isEmpty && !value.isEmpty }
-    Payload(nvPairs ++ filtered)
-  }
+  private[scalatracker] implicit class PayloadSyntax(payload: Payload) {
```

Comment:
```suggestion
  private[scalatracker] implicit class PayloadSyntax(payload: Payload) extends AnyVal {
```

In fact, `AnyVal` is so bad in doing what it promises (avoiding unnecessary boxing) that the only place where it's 100% safe is `Ops` classes.

Reply:
This is not immediately possible, because of this compiler error:
> value class may not wrap another user-defined value class

It's because Payload already extends AnyVal.

I see 3 options:
1. leave it as it is.
2. Remove the `extends AnyVal` from Payload and put it instead on PayloadSyntax
3. Remove the PayloadSyntax class entirely and add these methods directly as properties of the Payload class.

Hmmm... option 3 is probably the best suggestion.  wdyt @chuwy ?

Reply:
Ah, ok sorry - I didn't realise `Payload` is `AnyVal` as well. My gut feeling tells me we'll have less boxings if we use `AnyVal` for this class rather than for `Payload`, but 1) I have absolutely no proofs; 2) Even in that case it will be just a micro-optimisation.

Marking as solved.